packages/components/src/progressive-disclosure-panel/index.js

```diff
@@ -0,0 +1,137 @@
+	const filteredChildren = useMemo( () => {
+		return Array.isArray( children ) ? children.filter( isMenuItem ) : [];
+	}, [ children ] );
+
+	// Refresh which children should be reflected in the menu and what their
+	// associated menu item's state is; checked or not.
+	useEffect( () => {
+		const items = {};
+
+		filteredChildren.forEach( ( { props: childProps } ) => {
+			const { hasValue, isShownByDefault, label } = childProps;
+
+			// Menu item is checked if:
+			// - it currently has a value
+			// - or it was checked in previous menuItems state.
+			const isChecked = hasValue() || menuItems[ label ] === true;
+
+			// Menu item will be `disabled` if:
+			// - it is not checked
+			// - and is shown by default.
+			const isDisabled = ! isChecked && isShownByDefault;
+
+			items[ label ] = isDisabled ? 'disabled' : isChecked;
+		} );
+
+		setMenuItems( items );
+	}, [ filteredChildren ] );
+
+	if ( filteredChildren.length === 0 ) {
+		return null;
+	}
+
+	const getChildByMenuLabel = ( label ) => {
+		return filteredChildren.find(
+			( child ) => child.props.label === label
+		);
+	};
+
+	// Toggles the customized state of the child and its display if it isn't to
+	// be displayed by default. When toggling a child it's callback is executed.
+	const toggleChild = ( label ) => {
+		const wasSelected = menuItems[ label ];
+		const child = getChildByMenuLabel( label );
+		const { onDeselect, onSelect, isShownByDefault } = child.props;
+
+		if ( wasSelected && onDeselect ) {
+			onDeselect();
+		}
+
+		if ( ! wasSelected && onSelect ) {
+			onSelect();
+		}
+
+		// If child is was checked but is no longer and also shown by default
+		// disable the child's menu item.
+		const isDisabled = wasSelected && isShownByDefault;
+
+		setMenuItems( {
+			...menuItems,
+			[ label ]: isDisabled ? 'disabled' : ! wasSelected,
+		} );
+	};
+
+	// Resets display of children and executes resetAll callback if available.
+	const resetAllChildren = () => {
+		if ( typeof resetAll === 'function' ) {
+			resetAll();
+		}
+
+		// Turn off all menu items. Default controls will continue to display
+		// by virtue of their `isShownByDefault` prop however their menu item
+		// will be disabled to prevent behaviour where toggling has no effect.
+		const resetMenuItems = {};
+
+		filteredChildren.forEach( ( { props: childProps } ) => {
+			const { label, isShownByDefault } = childProps;
+			resetMenuItems[ label ] = isShownByDefault ? 'disabled' : false;
+		} );
+
+		setMenuItems( resetMenuItems );
+	};
+
+	const classes = classnames(
+		'components-progressive-disclosure-panel',
+		className
+	);
+
+	return (
+		<div className={ classes }>
+			<PanelContext.Provider value={ menuItems }>
+				<ProgressiveDisclosurePanelTitle
+					menuLabel={ menuLabel }
+					title={ title }
+					toggleChild={ toggleChild }
+					resetAll={ resetAllChildren }
+				/>
+				{ children }
+			</PanelContext.Provider>
+		</div>
```

Comment:
I have a strong preference to _not_ introducing new Sass based components that will later need to be migrated to Emotion.

Likewise with TypeScript. There is a goal of typing the entire components package as part of #18838. Given the trouble you've already gone to document the types for the component, we should go ahead and add TypeScript as well (preferrably by using native TypeScript, i.e., rename the files to `ts(x)`) and then follow the patterns in other components.

Likewise, to follow the pattern of new components being introduced into the component library, we should also make this a PolymorphicComponent that is connected to the new context system, using `useContextSystem` and `contextConnect` (there are many examples of this throughout the component library now).